### ISSUE TITLE
Even though the index_parallel task is marked failed, many of its single_phase_sub_tasks are still running.

### ISSUE THREAD

**Issue by didip** [apache/druid#11396]
### Affected Version

Tested on 0.21.1

### Description

Please include as much detailed information about the problem as possible.
- 15 middle managers with 20 workers each.
- We are deploying Druid inside Kubernetes.
- Each middle manager pods has 32GB RAM and 20 CPU.
- The configuration of the cluster is pretty basic, we don't use any affinity stuff.
- The native ingestion job uses maxNumSegmentsToMerge=100. The input data is around 3TB per day with hundreds of parquet files.
- To reproduce, we just keep it running for almost a day, we will see that index_parallel is marked failed but many of the subtasks are still running.

**Comment by jihoonson** [comment]
Hi @didip, thank you for your report. I have a couple of questions.

- Are those all remaining subtasks created by the `index_parallel` task that is marked failed? All subtasks should have the same group ID as their `index_parallel` task.
- The `index_parallel` task should clean up all running subtasks when it fails. If it didn't, there should be something bad happened before it cleaned up. Do you see anything interesting in the logs of the `index_parallel` task?

**Comment by didip** [comment]
Yes, all the remaining subtasks are created by the parent index_parallel.

Because I deployed Druid on Kubernetes, it is possible that some middle manager pods are restarted. When that happened I lost access to the temporary logs.

**Comment by jihoonson** [comment]
Hmm, I suppose that it's also possible that those middle manager pods were restarted while the parallel task was canceling its subtasks after it fails. Or, the parallel task failed because of the restart of the pods. Druid currently doesn't clean up subtasks in these cases. Ideally, the overlord should cancel those remaining subtasks.

**Comment by didip** [comment]
Yes, ideally Overlord should have bookkeeping on these and cancels the remaining subtasks.

**Comment by jihoonson** [comment]
> In a dynamic cloud like Kubernetes, hosts come and go, it would be really nice if tasks can recover or be cleaned up correctly upon restart.

Good idea. I think we should do both :slightly_smiling_face:
